File: client-clojure/src/sirqul_io_t_platform/specs/shipment_batch.clj
(ns sirqul-io-t-platform.specs.shipment-batch
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.retailer :refer :all]
            [sirqul-io-t-platform.specs.service-hub :refer :all]
            [sirqul-io-t-platform.specs.program :refer :all]
            [sirqul-io-t-platform.specs.shipment-import-status :refer :all]
            [sirqul-io-t-platform.specs.shipment :refer :all]
            )
  (:import (java.io File)))


(def shipment-batch-data
  {
   (ds/opt :id) int?
   (ds/opt :created) inst?
   (ds/opt :updated) inst?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :retailer) retailer-spec
   (ds/opt :hub) service-hub-spec
   (ds/opt :program) program-spec
   (ds/opt :shipmentImportStatuses) (s/coll-of shipment-import-status-spec)
   (ds/opt :shipments) (s/coll-of shipment-spec)
   (ds/opt :uploaded) inst?
   (ds/opt :started) inst?
   (ds/opt :optimizationStarted) inst?
   (ds/opt :completed) inst?
   (ds/opt :optimizationCompleted) inst?
   (ds/opt :source) string?
   (ds/opt :externalId) string?
   (ds/opt :total) int?
   (ds/opt :processedCount) int?
   (ds/opt :successCount) int?
   (ds/opt :failedCount) int?
   (ds/opt :validCount) int?
   (ds/opt :invalidCount) int?
   (ds/opt :averageTravelSpeed) float?
   (ds/opt :maxSeconds) int?
   (ds/opt :importStatus) string?
   (ds/opt :totalRoutes) int?
   (ds/opt :routesCreated) int?
   (ds/opt :totalLegsRequests) int?
   (ds/opt :legsRequested) int?
   (ds/opt :initScore) int?
   (ds/opt :hardScore) int?
   (ds/opt :mediumScore) int?
   (ds/opt :softScore) int?
   })

(def shipment-batch-spec
  (ds/spec
    {:name ::shipment-batch
     :spec shipment-batch-data}))

File: client-scalaz/src/main/scala/org/openapitools/client/api/ShipmentBatch.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import ShipmentBatch._

case class ShipmentBatch (
  id: Option[Long],
created: Option[OffsetDateTime],
updated: Option[OffsetDateTime],
active: Option[Boolean],
valid: Option[Boolean],
retailer: Option[Retailer],
hub: Option[ServiceHub],
program: Option[Program],
shipmentImportStatuses: Option[List[ShipmentImportStatus]],
shipments: Option[List[Shipment]],
uploaded: Option[OffsetDateTime],
started: Option[OffsetDateTime],
optimizationStarted: Option[OffsetDateTime],
completed: Option[OffsetDateTime],
optimizationCompleted: Option[OffsetDateTime],
source: Option[String],
externalId: Option[String],
total: Option[Integer],
processedCount: Option[Integer],
successCount: Option[Integer],
failedCount: Option[Integer],
validCount: Option[Integer],
invalidCount: Option[Integer],
averageTravelSpeed: Option[Double],
maxSeconds: Option[Long],
importStatus: Option[ImportStatus],
totalRoutes: Option[Integer],
routesCreated: Option[Integer],
totalLegsRequests: Option[Integer],
legsRequested: Option[Integer],
initScore: Option[Integer],
hardScore: Option[Long],
mediumScore: Option[Long],
softScore: Option[Long])

object ShipmentBatch {
  import DateTimeCodecs._
  sealed trait ImportStatus
  case object `NEW` extends ImportStatus
  case object STARTED extends ImportStatus
  case object FETCHINGDISTANCEDATA extends ImportStatus
  case object OPTIMIZING extends ImportStatus
  case object CREATINGROUTES extends ImportStatus
  case object COMPLETED extends ImportStatus
  case object INCOMPLETE extends ImportStatus

  object ImportStatus {
    def toImportStatus(s: String): Option[ImportStatus] = s match {
      case "`NEW`" => Some(`NEW`)
      case "STARTED" => Some(STARTED)
      case "FETCHINGDISTANCEDATA" => Some(FETCHINGDISTANCEDATA)
      case "OPTIMIZING" => Some(OPTIMIZING)
      case "CREATINGROUTES" => Some(CREATINGROUTES)
      case "COMPLETED" => Some(COMPLETED)
      case "INCOMPLETE" => Some(INCOMPLETE)
      case _ => None
    }

    def fromImportStatus(x: ImportStatus): String = x match {
      case `NEW` => "`NEW`"
      case STARTED => "STARTED"
      case FETCHINGDISTANCEDATA => "FETCHINGDISTANCEDATA"
      case OPTIMIZING => "OPTIMIZING"
      case CREATINGROUTES => "CREATINGROUTES"
      case COMPLETED => "COMPLETED"
      case INCOMPLETE => "INCOMPLETE"
    }
  }

  implicit val ImportStatusEnumEncoder: EncodeJson[ImportStatus] =
    EncodeJson[ImportStatus](is => StringEncodeJson(ImportStatus.fromImportStatus(is)))

  implicit val ImportStatusEnumDecoder: DecodeJson[ImportStatus] =
    DecodeJson.optionDecoder[ImportStatus](n => n.string.flatMap(jStr => ImportStatus.toImportStatus(jStr)), "ImportStatus failed to de-serialize")

  implicit val ShipmentBatchCodecJson: CodecJson[ShipmentBatch] = CodecJson.derive[ShipmentBatch]
  implicit val ShipmentBatchDecoder: EntityDecoder[ShipmentBatch] = jsonOf[ShipmentBatch]
  implicit val ShipmentBatchEncoder: EntityEncoder[ShipmentBatch] = jsonEncoderOf[ShipmentBatch]
}
